modules/optimization/test/tests/vectorpostprocessors/element_reaction_inner_product/element_reaction_inner_product.i

```diff
@@ -0,0 +1,128 @@
+[Mesh]
+  [gmg]
+    type = GeneratedMeshGenerator
+    dim = 1
+    nx = 10
+  []
+[]
+
+[AuxVariables]
+  [u]
+  []
+  [v]
+    initial_condition = 1
+  []
+  [dp1]
+  []
+  [dp2]
+  []
+[]
+
+p1 = 3.14
+p2 = 2.72
+[Reporters]
+  [params]
+    type = ConstantReporter
+    real_vector_names = 'p'
+    real_vector_values = '${p1} ${p2}'
+  []
+[]
+
+[Functions]
+  [p_fun]
+    type = ParsedOptimizationFunction
+    expression = 'p_1*x + p_2*p_2'
+    param_symbol_names = 'p_1 p_2'
+    param_vector_name = params/p
+  []
+  [u_fun]
+    type = ParsedFunction
+    expression = 'x'
+  []
+  [dp1_fun]
+    type = ParsedFunction
+    expression = 'x'
+    symbol_names = 'p_1 p_2'
+    symbol_values = '${p1} ${p2}'
+  []
+  [dp2_fun]
+    type = ParsedFunction
+    expression = '2*p_2'
+    symbol_names = 'p_1 p_2'
+    symbol_values = '${p1} ${p2}'
+  []
+[]
+
+[ICs]
+  [u_ic]
+    type = FunctionIC
+    variable = u
+    function = u_fun
+  []
+  [dp1_ic]
+    type = FunctionIC
+    variable = dp1
+    function = dp1_fun
+  []
+  [dp2_ic]
+    type = FunctionIC
+    variable = dp2
+    function = dp2_fun
+  []
+[]
+
+[VectorPostprocessors]
+  [inner_product]
+    type = ElementOptimizationReactionFunctionInnerProduct
```

Comment:
you should just make a ParsedVPP a this point

Reply:
So far, all of the gradients have been computed in a different enough way that it can't be put into a parsed function.  They depend on the weak function of the kernel.  This object takes inner products of the variables; ElementOptimizationDiffusionCoefFunctionInnerProduct takes an inner product of the gradients of the variables.  We haven't found an easy way to generalize that.